pkg/controller/core/reconciler.go

```diff
@@ -0,0 +1,261 @@
+	if ok && skip == "true" {
+		log.Info("Skipping reconcile due to annotation")
+		return ctrl.Result{}, nil
+	}
+
+	// build context for components
+	compLog := log.WithName("components")
+	ctx := &Context{
+		Context:  rootCtx,
+		Object:   obj,
+		Client:   r.client,
+		Patch:    r.patcher,
+		Scheme:   r.mgr.GetScheme(),
+		Recorder: r.recorder,
+	}
+
+	// reconcile components
+	var finalRes ctrl.Result
+	var errs []error
+	for _, rc := range r.components {
+		res := ctrl.Result{}
+		var err error
+
+		ctx.Log = compLog.WithName(rc.name)
+
+		if ctx.Object.GetDeletionTimestamp().IsZero() {
+			log.Info("Reconciling component", "component", rc.name)
+			res, err = rc.comp.Reconcile(ctx)
+
+			if rc.finalizer != nil && !controllerutil.ContainsFinalizer(ctx.Object, rc.finalizerName) {
+				log.Info("Registering finalizer", "component", rc.name)
+				controllerutil.AddFinalizer(ctx.Object, rc.finalizerName)
+			}
+		} else if rc.finalizer != nil && controllerutil.ContainsFinalizer(ctx.Object, rc.finalizerName) {
+			log.Info("Finalizing component", "component", rc.name)
+
+			var done bool
+			res, done, err = rc.finalizer.Finalize(ctx)
+			if done {
+				log.Info("Removing finalizer", "component", rc.name)
+				controllerutil.RemoveFinalizer(ctx.Object, rc.finalizerName)
+			}
+		}
+
+		if res.Requeue {
+			finalRes.Requeue = true
+		}
+		if res.RequeueAfter != 0 && (finalRes.RequeueAfter == 0 || finalRes.RequeueAfter > res.RequeueAfter) {
+			finalRes.RequeueAfter = res.RequeueAfter
+		}
+		if err != nil {
+			log.Error(err, "Component reconciliation failed", "component", rc.name)
+			errs = append(errs, err)
+		}
+	}
+
+	// patch metadata when changes occur
+	currentMeta := r.apiType.DeepCopyObject().(client.Object)
+	currentMeta.SetName(ctx.Object.GetName())
+	currentMeta.SetNamespace(ctx.Object.GetNamespace())
+	currentMeta.SetLabels(ctx.Object.GetLabels())
+	currentMeta.SetAnnotations(ctx.Object.GetAnnotations())
+	currentMeta.SetFinalizers(ctx.Object.GetFinalizers())
+
+	cleanMeta := r.apiType.DeepCopyObject().(client.Object)
+	cleanMeta.SetName(cleanObj.GetName())
+	cleanMeta.SetNamespace(cleanObj.GetNamespace())
+	cleanMeta.SetLabels(cleanObj.GetLabels())
+	cleanMeta.SetAnnotations(cleanObj.GetAnnotations())
+	cleanMeta.SetFinalizers(cleanObj.GetFinalizers())
+
+	patch := client.MergeFrom(cleanMeta)
+	json, _ := patch.Data(currentMeta)
+
+	log.V(1).Info("Patching metadata", "type", patch.Type(), "patch", string(json))
+	if err := r.client.Patch(ctx, currentMeta, patch, &client.PatchOptions{FieldManager: r.name}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("error patching metadata: %w", err)
+	}
+
+	// condense all error messages into one
+	var err error
+	if len(errs) == 1 {
+		err = errs[0]
+	} else if len(errs) > 1 {
+		sb := strings.Builder{}
+		for _, e := range errs {
+			sb.WriteString(fmt.Sprintf(" %s\n", e.Error()))
+		}
+		err = errors.New(sb.String())
+	}
+
+	return finalRes, err
```

Comment:
What do you think of https://github.com/kubernetes/apimachinery/blob/30802a09abe3f7b38aa5749c1ec5f6443a85e6ed/pkg/util/errors/errors.go#L35?

Reply:
I think it's better to use an existing type like this instead of writing my own. 😛 Nice find.